Code review recommended practices (#727)
microsoft/code-with-engineering-playbook: docs/code-reviews/process-guidance/README.md

1. Assess the time to merge across sprints to see if the process is improving.
1. Ping required approvers directly as a reminder.

## Recommendations for a better code review

### Take your time to review but don't go beyond of the specific task

You shouldn't review code too quickly but neither take too long in one sitting. If you have many pull requests or if the complexity of code is demanding the recommendation is to take a break between reviews to recover, and focus on the ones you are most experienced with.

Always remember, the goal of a code review is to verify that the specified task has been achieved. If your suggestions are on code related to, but not included in the PR, raise those suggestions as separate tasks (bugs, technical debt etc.). Don't block the current PR on suggestions outside the scope of the PR.

> ### Automate whenever possible

Reviewer: There is already a section on automation in the reviewer guidance (under "Decide on a common standard for each language"). Rather than duplicating this - can we consolidate the information in these sections? I personally think that the consolidated section fits better here than under the reviewer section (as it is general preparation the team needs to do - but we don't want to have duplicate sections)

1. Assess the time to merge across sprints to see if the process is improving.
1. Ping required approvers directly as a reminder.

## Recommendations for a better code review

### Get a process for fixing found issues

Before getting started with any commit, your team should define how you will measure the effectiveness of peer review and get some goals. Using a [SMART criteria](https://en.wikipedia.org/wiki/SMART_criteria) the team can get this objectives in the most accurate way and not something very general like **fix more bugs**, the more precise a goal can be the easier it will be to hit the target.

> Metrics can play a significant role in this part, you can include an **Inspection rate**(the speed with which review is performed), **Defect rate**(the number of bugs found within an hour of review) and **Defect density**(the number of bugs found per piece of code) as the most common ones to be considered.

Reviewer: ```suggestion
Metrics can play a significant role in this part, you can include an **Inspection rate** (the speed with which review is performed), **Defect rate** (the number of bugs found within an hour of review) and **Defect density** (the number of bugs found per piece of code) as the most common ones to be considered.
```

> ### Code reviews shouldn't include many lines of code

Reviewer: This sounds like an author specific guidance, can we move it there?

Author: I took the section to the author guidance.

It's easy to say a developer can review 400 or more lines but when code surpasses certain amount of lines of code effectiveness of defects discovery can decrease and there is a chance of not having a good review or just letting bugs pass to main branch or another stage. Setting a line of code limit is as important as a time limit.

> ### Take your time to review but don't go beyond of the specific task

Reviewer: This sounds like a reviewer specific guidance, can we move it to the reviewer section?

Reviewer: This was just resolved without action... was there a reason?

Author: It's already there.

> There is a small, very small gap when reviewing code talking about time, you shouldn't review code too quickly but neither take too long in one sitting. If you have many pull requests that could lead to many code reviews or complexity of code is demanding the recommendation is to not invest much time, take a break, spend some minutes between reviews to recover and focus in the ones you are most experienced.

Reviewer: ```suggestion
You shouldn't review code too quickly but neither take too long in one sitting. If you have many pull requests or if the complexity of code is demanding the recommendation is to take a break between reviews to recover, and focus one the ones you are most experienced with.
```

> Always remember, the goal in a code review is to verify the specified task has been achieved, there is a common practice of analyzing tasks beyond the scope of a PR by suggesting improvements that are are not being considered for this you can raise a bug, an improvement or a technical debt but not postpone a review because of a different task.

Reviewer: ```suggestion
Always remember, the goal of a code review is to verify that the specified task has been achieved. If your suggestions are on code related to, but not included in the PR, raise those suggestions as separate tasks (bugs, technical debt etc.). Don't block the current PR on suggestions outside the scope of the PR.
```

### Automate whenever possible

> Before doing any code review it's important to set automated builds, tests and checks (something achievable in the [CI process](../../continuous-integration/README.md)), this can save human reviewers some time and focus in areas like design and functionality for proper evaluation. Reviewing code with some questions in mind can help team focus on the right things. One of the possible questions to ask when reviewing is **Do I understand what the code does?** and it's important to start from there in order to get the best qualified team reviewing the code.

Reviewer: ```suggestion
Before doing any code review it's important to set automated builds, tests and checks (something achievable in the [CI process](../../continuous-integration/README.md)), this can save human reviewers some time and focus in areas like design and functionality for proper evaluation.
```

Reviewer: The last part was not related to the topic "automate whenever possible" and is also covered in the reviewer guidance already

Author: Thanks I saw it already.

> This way of critical evaluation will ensure the team is checking the right things and it will reduce time when it comes to deal with human reviewing.

Reviewer: ```suggestion
This will ensure the team is focusing on the right things and it will reduce review time.
```
